rfcs/0042-networking-refactor.md

```diff
@@ -0,0 +1,337 @@
+
+Here is a list of the IPC messages that will be supported in each direction, along with some relevant type definitions in Go:
+
+```txt
+Daemon -> Helper
+  // it's possible to just remove this message as it is just a specialized case of `sendRequests`
+  sendRequestToPeer(requestId RequestId, to Peer, msgType MsgType, rawData []byte)
+  sendRequests(requestId RequestId, to AbstractPeerGraph, msgType MsgType, rawData []byte)
+  sendResponse(requestId RequestId, status ValidationStatus, rawData []byte)
+  broadcast(msgType MsgType, rawData []byte)
+  validate(validation ValidationHandle, status ValidationStatus)
+
+Helper -> Daemon
+  handleRequest(requestId RequestId, from Peer, rawData []byte)
+  handleResponse(requestId RequestId, validation ValidationHandle, rawData []byte)
+  handleGossip(from Peer, validation ValidationHandle, rawData []byte)
+  stats(stats Stats)
+```
+
+```go
+// == The `Config` struct is sent with the `init` message at the start of the protocol.
+
+// The following old fields have been completely removed:
+//   - `metricsPort` (moving over to push-based stats syncing, where we will sync any metrics we want to expose)
+//   - `unsafeNotTrustIp` (only used as a hack in old integration test framework; having it makes p2p code harder to reason about)
+//   - `gaterConfig` (moving towards more abstracted interface in which Go manages gating state data)
+type Config struct {
+  networkId           string      // unique network identifier
+  privateKey          string      // libp2p id private key
+  stateDirectory      string      // directory to store state in (peerstore and dht will be stored/loaded from here)
+  listenOn            []Multiaddr // interfaces we listen on
+  externalAddr        Multiaddr   // interface we advertise for other nodes to connect to
+  floodGossip         bool        // enables gossip flooding (should only be turned on for protected nodes hidden behind a sentry node)
+  directPeers         []Multiaddr // forces the node to maintain connections with peers in this list (typically only used for sentry node setups and other specific networking scenarios; these peers are automatically trustlisted)
+  seedPeers           []Multiaddr // list of seed peers to connect to initially (seeds are automatically trustlisted)
+  maxConnections      int         // maximum number of connections allowed before the connection manager begins trimming open connections
+  validationQueueSize int         // size of the queue of active pending validation messages
+  // TODO: peerExchange bool vs minaPeerExchange bool (seems like at least one of these should be deprecated)
```

Comment:
It's probably worth trying to see when we do the refactor if the libp2p builtin peerExchange starts working (maybe we did something weird that disabled it in our current implementation) before falling back to our custom peer exchange